### PR TITLE
Swati/join split circuit

### LINKED ISSUE #45: Add join/split circuits for partitioned commitments (?)
One known issue with partitioned states (used in ZCash, Nightfall, etc.) is that many commitments representing one state can reduce to 'dust'. i.e. the values get smaller and smaller as we transfer change back to the owner. 
Starlight currently assumes you manage your commitments so this either doesn't happen or there is some transfer function allowing you to transfer values to yourself (like a 'join' but with extra steps).
Perhaps we should add join/split circuits and shield contract functions to make this easier for users?

### PR DESCRIPTION
**Pull request by SwatiEY**
close #45 
join Commitments circuit is added to the circuits, whenever we have a partitioned state and it is called when we don't have two appropriate commitments.

### CODE REVIEW

**Comment by MirandaWood** [comment]
Can we make it so the `joinCommitments` code is only added to the new files + setup command if we have partitioned states? Currently a lot of it is if `nullifiersRequired` which would also add it for whole states. Thank you!

File: src/boilerplate/orchestration/javascript/raw/boilerplate-generator.ts [hunk 1]
```diff
@@ -111,20 +112,44 @@ class BoilerplateGenerator {
...
+            \n let ${stateName}_witness_1 ;
+
+                        while( commitmentFlag === false) {
+                console.log('Existing Commitments are not appropriate and we need to call Join Circuit');
+                \n  ${stateName}_witness_0 = await getMembershipWitness('${contractName}', generalise(${stateName}_0_oldCommitment).integer);
+                \n  ${stateName}_witness_1 = await getMembershipWitness('${contractName}', generalise(${stateName}_1_oldCommitment).integer);
+
+                const tx = await joinCommitments('${contractName}', '${mappingName}${mappingKey}', secretKey, publicKey, ${stateName}_stateVarId, ${stateName}_preimage, [${stateName}_0_oldCommitment,${stateName}_1_oldCommitment], [${stateName}_witness_0,${stateName}_witness_1], instance);
+
+                console.log(tx);
```
Comment (MirandaWood) [inline src/boilerplate/orchestration/javascript/raw/boilerplate-generator.ts:135]: could we remove this log, or replace it with something like `your commitments are x`? Just as the tx is a big ugly log :)

File: src/boilerplate/orchestration/javascript/raw/boilerplate-generator.ts [hunk 2]
```diff
@@ -251,6 +280,7 @@ class BoilerplateGenerator {
         `\nimport fs from 'fs';
         \n`,
         `\nimport { getContractInstance, registerKey, getInputCommitments } from './common/contract.mjs';`,
+        `\nimport  joinCommitments  from './common/contract.mjs';`,
```
Comment (MirandaWood) [inline src/boilerplate/orchestration/javascript/raw/boilerplate-generator.ts:283]: Just for cleanliness, we could join together these two lines (if the imports are unused, that's fine):
```suggestion
        `\nimport { getContractInstance, registerKey, getInputCommitments, joinCommitments } from './common/contract.mjs';`,
```

File: src/boilerplate/contract/solidity/raw/ContractBoilerplateGenerator.ts [hunk 3]
```diff
@@ -221,16 +223,56 @@ class ContractBoilerplateGenerator {
         ),
       ];
 
+      if (isjoinCommitmentsFunction?.includes('true')) {
+
+       joinCommitmentsInputs.push(
+        `
+           function joinCommitments(uint256[] calldata newNullifiers, uint256 commitmentRoot, uint256[] calldata newCommitments, uint256[] calldata proof) public {
+
+            bytes4 sig = bytes4(keccak256("joinCommitments(uint256[],uint256,uint256[],uint256[])"));
```
[the remark this answers is not in the document]
Comment (SwatiEY) [reply src/boilerplate/contract/solidity/raw/ContractBoilerplateGenerator.ts:232]: removed it..thanks

File: src/boilerplate/orchestration/javascript/raw/boilerplate-generator.ts (hunk 2 above)
Comment (SwatiEY) [reply src/boilerplate/orchestration/javascript/raw/boilerplate-generator.ts]: thanks for the suggestion.. combined the import call

**Review by MirandaWood — APPROVED** [review]

**Review by dwebchapey — APPROVED** [review]

**Pull request merged**